pennylane/qnodes/base.py

```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
                 )
             self.queue.append(op)  # TODO rename self.queue to self.op_queue
 
+    def _determine_structured_variable_name(self, parameter_value, prefix):
+        """Determine the variable names corresponding to a parameter.
+
+        This method unrolls the parameter value if it has an array
+        or list structure.
+
+        Args:
+            parameter_value (Union[Number, Sequence[Any], array[Any]]): The value of the parameter. This will be used as a blueprint for the returned variable name(s).
+            prefix (str): Prefix that will be added to the variable name(s), usually the parameter name
+
+        Returns:
+            Union[str,Sequence[str],array[str]]: The variable name(s) in the same structure as the parameter value
+        """
+        if isinstance(parameter_value, np.ndarray):
+            variable_name_string = np.empty_like(parameter_value, dtype=object)
+
+            for index in np.ndindex(*variable_name_string.shape):
+                variable_name_string[index] = "{}[{}]".format(prefix, ",".join([str(i) for i in index]))
+        elif isinstance(parameter_value, Sequence):
+            variable_name_string = []
+
+            for idx, val in enumerate(parameter_value):
+                variable_name_string.append(self._determine_structured_variable_name(val, "{}[{}]".format(prefix, idx)))
+        else:
+            variable_name_string = prefix
+
+        return variable_name_string
+
+    def _make_variables(self, args, kwargs):
+        """Create the :class:`~.variable.Variable` instances representing the QNode's arguments.
+
+        The created :class:`~.variable.Variable` instances are given in the same nested structure
+        as the original arguments. The :class:`~.variable.Variable` instances are named according
+        to the python expression that adresses them inside the qfunc. Consider the following example:
+
+        .. code-block:: python
+            @qml.qnode(dev)
+            def qfunc(a, w):
+                qml.Hadamard(0)
+                qml.CRX(a, wires=[0, 1])
+                qml.Rot(w[0], w[1], w[2], wires=[1])
+                qml.CRX(-a, wires=[0, 1])
+
+                return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+            qfunc(3.4, [1.2, 3.4, 5.6])
```

Comment:
Might look cleaner to move this below to the 'prompt' codeblock?

Reply:
Yeah, you're right!